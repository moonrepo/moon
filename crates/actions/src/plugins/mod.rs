mod commands;
mod operations;

pub use commands::*;
pub use operations::*;
