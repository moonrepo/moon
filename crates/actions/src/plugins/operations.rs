use convert_case::{Case, Casing};
use moon_action::{Action, ActionStatus, Operation};
use moon_common::Id;
use moon_pdk_api::{Operation as PluginOperation, OperationStatus, SyncOutput, VirtualPath};
use moon_time::chrono::{DateTime, Local};
use moon_toolchain_plugin::{CallResult, ToolchainPlugin};

pub fn convert_plugin_operation(
    toolchain: &ToolchainPlugin,
    base: PluginOperation,
) -> miette::Result<Operation> {
    let mut op = Operation::sync_operation(base.id.to_case(Case::Kebab))?;

    op.plugin = Some(Id::new(&toolchain.id)?);

    op.started_at = base
        .started_at
        .map(|ts| DateTime::<Local>::from(ts).naive_utc())
        .unwrap();

    op.finished_at = base
        .finished_at
        .map(|ts| DateTime::<Local>::from(ts).naive_utc());

    op.duration = base.duration;

    op.status = match base.status {
        OperationStatus::Pending => ActionStatus::Running,
        OperationStatus::Failed => ActionStatus::Failed,
        OperationStatus::Passed => ActionStatus::Passed,
    };

    Ok(op)
}

pub fn convert_plugin_operations(
    toolchain: &ToolchainPlugin,
    base: Vec<PluginOperation>,
) -> miette::Result<Vec<Operation>> {
    let mut ops = vec![];

    for item in base {
        ops.push(convert_plugin_operation(toolchain, item)?);
    }

    Ok(ops)
}

pub fn inherit_changed_files(op: &mut Operation, files: Vec<VirtualPath>) {
    if let Some(meta) = op.get_file_state_mut() {
        for file in files {
            if let Some(file) = file.real_path() {
                meta.changed_files.push(file);
            }
        }
    }
}

pub fn finalize_action_operations(
    action: &mut Action,
    toolchain: &ToolchainPlugin,
    mut op: Operation,
    plugin_ops: Vec<PluginOperation>,
    changed_files: Vec<VirtualPath>,
) -> miette::Result<()> {
    op.plugin = Some(Id::new(&toolchain.id)?);

    if op.status == ActionStatus::Running {
        op.finish(ActionStatus::Passed);
    }

    // Inherit plugin operations
    action
        .operations
        .extend(convert_plugin_operations(toolchain, plugin_ops)?);

    // Inherit changed files
    inherit_changed_files(&mut op, changed_files);

    action.operations.push(op);

    Ok(())
}

pub fn finalize_sync_operation(sync_result: CallResult<SyncOutput>) -> miette::Result<Operation> {
    // Add an operation for the overall sync
    let mut op = convert_plugin_operation(&sync_result.toolchain, sync_result.operation)?;

    // Inherit plugin operations
    op.operations.extend(convert_plugin_operations(
        &sync_result.toolchain,
        sync_result.output.operations,
    )?);

    // Inherit changed files
    inherit_changed_files(&mut op, sync_result.output.changed_files);

    Ok(op)
}
