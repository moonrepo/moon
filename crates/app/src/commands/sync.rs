use super::syncs::codeowners::SyncCodeownersArgs;
use super::syncs::config_schemas::SyncConfigSchemasArgs;
use super::syncs::hooks::SyncHooksArgs;
use crate::components::run_action_pipeline;
use crate::session::MoonSession;
use clap::Subcommand;
use iocraft::prelude::element;
use moon_action_graph::ActionGraphBuilderOptions;
use moon_console::ui::{Container, Notice, StyledText, Variant};
use starbase::AppResult;

#[derive(Clone, Debug, Subcommand)]
pub enum SyncCommands {
    #[command(
        name = "codeowners",
        about = "Aggregate and sync code owners to a `CODEOWNERS` file."
    )]
    Codeowners(SyncCodeownersArgs),

    #[command(
        name = "config-schemas",
        about = "Generate and sync configuration JSON schemas for use within editors."
    )]
    ConfigSchemas(SyncConfigSchemasArgs),

    #[command(
        name = "hooks",
        about = "Generate and sync hook scripts for the workspace configured VCS."
    )]
    Hooks(SyncHooksArgs),

    #[command(
        name = "projects",
        about = "Sync all projects and configs in the workspace."
    )]
    Projects,
}

pub async fn sync(session: MoonSession) -> AppResult {
    let workspace_graph = session.get_workspace_graph().await?;
    let mut action_graph_builder = session
        .build_action_graph_with_options(ActionGraphBuilderOptions::default())
        .await?;

    action_graph_builder.sync_workspace().await?;

    for project in workspace_graph.projects.get_all_unexpanded() {
        action_graph_builder.sync_project(project).await?;
    }

    let (action_context, action_graph) = action_graph_builder.build();

    run_action_pipeline(&session, action_context, action_graph).await?;

    session.console.render(element! {
        Container {
            Notice(variant: Variant::Success) {
                StyledText(content: format!("Synced the workspace and all projects"))
            }
        }
    })?;

    Ok(None)
}
