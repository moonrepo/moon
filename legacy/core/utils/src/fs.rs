use crate::time::{is_stale, to_millis};
use moon_common::consts::CONFIG_DIRNAME;
use starbase_utils::fs;
use std::path::{Path, PathBuf};
use std::time::Duration;

pub mod temp {
    use super::*;

    pub fn get_dir() -> PathBuf {
        crate::get_workspace_root()
            .join(CONFIG_DIRNAME)
            .join("cache")
            .join("temp")
    }

    pub fn get_file(source: &str, ext: &str) -> PathBuf {
        get_dir().join(format!("{}.{}", crate::hash(source), ext))
    }

    pub fn read<P: AsRef<Path>>(path: P) -> miette::Result<Option<String>> {
        let file = path.as_ref();

        if !file.exists() {
            return Ok(None);
        }

        if let Ok(metadata) = file.metadata()
            && let Ok(filetime) = metadata.created()
        {
            // Temp files only last for 4 hours (half a workday)
            if is_stale(to_millis(filetime), Duration::from_secs(60 * 60 * 4)) {
                fs::remove_file(file)?;

                return Ok(None);
            }
        }

        Ok(Some(fs::read_file(file)?))
    }

    pub fn write<P: AsRef<Path>, D: AsRef<str>>(path: P, data: D) -> miette::Result<()> {
        let path = path.as_ref();

        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write_file(path, data.as_ref())?;

        Ok(())
    }
}
