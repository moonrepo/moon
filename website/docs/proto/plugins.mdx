---
title: Plugins
---

proto supports a pluggable architecture as a means for consumers to integrate and manage custom
tools (languages, CLIs, etc) within proto's toolchain. It's not possible for proto to support
_everything_ in core directly, so plugins are a way for the community to extend the toolchain to
their needs.

## Enabling plugins

Plugins can be enabled by configuring them in [`.prototools`](./config#plugins) files, within the
`[plugins]` section. The map key is the plugin name _in kebab-case_, which is used as the
binary/tool name in proto, and also the name for configuration and cache purposes. The map value is
a [plugin locator string](../guides/wasm-plugins#configuring-plugin-locations) that defines a
protocol and source location.

```toml title=".prototools"
[plugins.tools]
<id> = "<protocol>://<location>"
```

## Creating plugins

To ease the plugin development process, proto supports 2 types of plugins, a
[non-WASM configuration based plugin](./non-wasm-plugin) for basic use cases, and a
[WASM based plugin](./wasm-plugin) for advanced use cases.

## Publish a plugin

proto's registry is currently powered by static JSON files located in our official
[proto repository](https://github.com/moonrepo/proto/tree/master/registry). View that link for
information on how to publish a plugin.
