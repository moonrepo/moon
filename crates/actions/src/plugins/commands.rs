use miette::IntoDiagnostic;
use moon_action::{ActionStatus, Operation};
use moon_app_context::AppContext;
use moon_args::join_args;
use moon_common::{
    Id, color,
    path::{PathExt, WorkspaceRelativePathBuf, encode_component},
};
use moon_console::{Checkpoint, Console};
use moon_env_var::GlobalEnvBag;
use moon_hash::hash_content;
use moon_pdk_api::{CacheInput, ExecCommand, ExecCommandInput, VirtualPath};
use moon_process::{Command, Output};
use moon_time::to_millis;
use moon_toolchain_plugin::ToolchainPlugin;
use starbase_utils::fs;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use tokio::task::JoinSet;
use tracing::{debug, warn};

hash_content!(
    struct ExecCommandHash<'data> {
        command: &'data ExecCommandInput,

        #[serde(skip_serializing_if = "BTreeMap::is_empty")]
        input_env: BTreeMap<String, String>,

        #[serde(skip_serializing_if = "BTreeMap::is_empty")]
        input_files: BTreeMap<WorkspaceRelativePathBuf, String>,
    }
);

pub type OnExecFn = Arc<dyn Fn(&ExecCommand, (u8, u8)) -> miette::Result<()> + Send + Sync>;

pub fn handle_on_exec(
    console: &Console,
    command: &ExecCommand,
    attempts: (u8, u8),
) -> miette::Result<()> {
    let input = &command.command;
    let label = command
        .label
        .clone()
        .unwrap_or_else(|| format!("{} {}", input.command, join_args(&input.args)));

    if attempts.0 > 1 {
        console.print_checkpoint_with_comments(
            Checkpoint::Setup,
            label,
            [format!("attempt {} of {}", attempts.0, attempts.1)],
        )
    } else {
        console.print_checkpoint(Checkpoint::Setup, label)
    }
}

#[derive(Clone, Default)]
pub struct ExecCommandOptions {
    pub on_exec: Option<OnExecFn>,
    pub prefix: String,
    pub working_dir: PathBuf,
}

async fn internal_exec_plugin_command(
    app_context: Arc<AppContext>,
    command: &ExecCommand,
    options: &ExecCommandOptions,
    attempts: (u8, u8),
) -> miette::Result<Output> {
    let input = &command.command;

    let mut cmd = Command::new(&input.command);
    cmd.args(&input.args);
    cmd.envs(&input.env);

    if let Some(cwd) = input.working_dir.as_ref().and_then(|dir| dir.real_path()) {
        cmd.cwd(cwd);
    } else {
        cmd.cwd(&options.working_dir);
    }

    cmd.with_console(app_context.console.clone());
    cmd.set_error_on_nonzero(!command.allow_failure);
    cmd.set_print_command(app_context.workspace_config.pipeline.log_running_command);

    // Must be last!
    app_context
        .toolchain_registry
        .prepare_process_command(&mut cmd);

    if let Some(on_exec) = &options.on_exec {
        on_exec(command, attempts)?;
    }

    if command.command.stream {
        cmd.exec_stream_output().await
    } else {
        cmd.exec_capture_output().await
    }
}

async fn internal_exec_plugin_command_as_operation(
    app_context: Arc<AppContext>,
    command: &ExecCommand,
    options: &ExecCommandOptions,
    attempts: (u8, u8),
) -> miette::Result<Operation> {
    let mut op = Operation::process_execution(&command.command.command); // hah

    let result = match &command.cache {
        Some(key) => {
            let mut hash_item = ExecCommandHash {
                command: &command.command,
                input_env: BTreeMap::new(),
                input_files: BTreeMap::new(),
            };

            if !command.inputs.is_empty() {
                gather_cache_inputs(&app_context, &command.inputs, &mut hash_item).await?;
            }

            app_context
                .clone()
                .cache_engine
                .execute_if_changed(
                    format!(
                        "{}:{}",
                        options.prefix,
                        encode_component(key).to_lowercase()
                    ),
                    hash_item,
                    async move |_| {
                        internal_exec_plugin_command(app_context, command, options, attempts).await
                    },
                )
                .await
        }
        None => internal_exec_plugin_command(app_context, command, options, attempts)
            .await
            .map(Some),
    };

    match result {
        Ok(maybe_output) => {
            if let Some(output) = maybe_output {
                op.finish_from_output(output.status(), output.stdout, output.stderr);
            } else {
                op.finish(ActionStatus::Skipped);
            }

            Ok(op)
        }
        Err(error) => {
            op.finish(ActionStatus::Failed);

            Err(error)
        }
    }
}

pub async fn exec_plugin_command(
    app_context: Arc<AppContext>,
    command: &ExecCommand,
    options: &ExecCommandOptions,
) -> miette::Result<Vec<Operation>> {
    let mut ops = vec![];
    let attempt_count = 1 + command.retry_count;

    for attempt in 1..=attempt_count {
        match internal_exec_plugin_command_as_operation(
            app_context.clone(),
            command,
            options,
            (attempt, attempt_count),
        )
        .await
        {
            Ok(op) => {
                let success = op
                    .get_exec_output()
                    .is_some_and(|exec| exec.get_exit_code() == 0);

                ops.push(op);

                if success {
                    break;
                }
            }
            Err(error) => {
                if attempt == attempt_count {
                    return Err(error);
                }
            }
        };

        debug!(
            "Failed to execute {} command, retrying...",
            color::shell(command.label.as_ref().unwrap_or(&command.command.command)),
        );
    }

    Ok(ops)
}

pub async fn exec_plugin_commands(
    toolchain: &ToolchainPlugin,
    app_context: Arc<AppContext>,
    commands: Vec<ExecCommand>,
    options: ExecCommandOptions,
) -> miette::Result<Vec<Operation>> {
    let mut serial = vec![];
    let mut parallel = vec![];
    let mut ops = vec![];

    for command in commands {
        if command.parallel {
            parallel.push(command);
        } else {
            serial.push(command);
        }
    }

    // Execute serial first, as a parallel command may
    // depend on a serial command having been executed
    for command in serial {
        ops.extend(exec_plugin_command(app_context.clone(), &command, &options).await?);
    }

    // Then execute the parallel commands
    let mut set = JoinSet::new();

    for command in parallel {
        let app_context = app_context.clone();
        let options = options.clone();

        set.spawn(async move { exec_plugin_command(app_context, &command, &options).await });
    }

    while let Some(result) = set.join_next().await {
        ops.extend(result.into_diagnostic()??);
    }

    // Inherit toolchain ID
    for op in &mut ops {
        op.plugin = Some(Id::new(&toolchain.id)?);
    }

    Ok(ops)
}

async fn gather_cache_inputs(
    app_context: &AppContext,
    inputs: &[CacheInput],
    hash_item: &mut ExecCommandHash<'_>,
) -> miette::Result<()> {
    let mut hashable_files = vec![];

    let get_file = |virtual_path: &VirtualPath,
                    workspace_root: &Path|
     -> Option<(PathBuf, WorkspaceRelativePathBuf)> {
        if let Some(abs_path) = virtual_path.real_path() {
            if let Ok(rel_path) = abs_path.relative_to(workspace_root) {
                if abs_path.exists() {
                    return Some((abs_path, rel_path));
                } else {
                    // Don't warn for missing files
                    return None;
                }
            }
        }

        warn!(
            path = virtual_path.to_string(),
            "Only real paths within the workspace can be used as a cache input, received an invalid virtual path",
        );

        None
    };

    for input in inputs {
        match input {
            CacheInput::EnvVar(name) => {
                hash_item.input_env.insert(
                    name.into(),
                    GlobalEnvBag::instance().get(name).unwrap_or_default(),
                );
            }
            CacheInput::FileHash(virtual_path) => {
                if let Some((_, rel_path)) = get_file(virtual_path, &app_context.workspace_root) {
                    hashable_files.push(rel_path);
                }
            }
            CacheInput::FileSize(virtual_path) | CacheInput::FileTimestamp(virtual_path) => {
                if let Some((abs_path, rel_path)) =
                    get_file(virtual_path, &app_context.workspace_root)
                {
                    let metadata = fs::metadata(&abs_path)?;

                    if matches!(input, CacheInput::FileSize(_)) {
                        hash_item
                            .input_files
                            .insert(rel_path, format!("size:{}", metadata.len()));
                    } else if let Ok(timestamp) =
                        metadata.modified().or_else(|_| metadata.created())
                    {
                        hash_item
                            .input_files
                            .insert(rel_path, format!("timestamp:{}", to_millis(timestamp)));
                    }
                }
            }
        };
    }

    if !hashable_files.is_empty() && app_context.vcs.is_enabled() {
        for (rel_path, hash) in app_context
            .vcs
            .get_file_hashes(&hashable_files, true)
            .await?
        {
            hash_item
                .input_files
                .insert(rel_path, format!("hash:{hash}"));
        }
    }

    Ok(())
}
