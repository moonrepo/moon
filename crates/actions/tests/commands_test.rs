use moon_action::{ActionStatus, Operation};
use moon_actions::plugins::*;
use moon_env_var::GlobalEnvBag;
use moon_pdk_api::{CacheInput, ExecCommand, ExecCommandInput, VirtualPath};
use moon_test_utils2::WorkspaceMocker;
use starbase_sandbox::{Sandbox, assert_snapshot, create_empty_sandbox};
use starbase_utils::json;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

fn create_workspace() -> (Sandbox, WorkspaceMocker) {
    let sandbox = create_empty_sandbox();
    let mocker = WorkspaceMocker::new(sandbox.path()).with_default_projects();

    (sandbox, mocker)
}

mod plugin_commands {
    use super::*;

    mod exec_one {
        use super::*;

        #[tokio::test]
        async fn handles_success_failure() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let ops = exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["success"])),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 0);
            assert_eq!(exec.stdout.as_deref().unwrap().trim(), "success");

            let ops = exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(ExecCommandInput::pipe("exit", ["3"])).allow_failure(),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 3);
        }

        #[tokio::test]
        #[should_panic(expected = "exit code 3")]
        async fn errors_for_nonzero() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(ExecCommandInput::pipe("exit", ["3"])),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();
        }

        #[tokio::test]
        async fn can_set_env() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let mut input = ExecCommandInput::pipe("echo", ["$TEST"]);
            input.env.insert("TEST".into(), "value".into());

            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(input),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 0);
            assert_eq!(exec.stdout.as_deref().unwrap().trim(), "value");
        }

        #[tokio::test]
        async fn can_set_cwd() {
            let (sandbox, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            // Current dir
            let mut input = ExecCommandInput::pipe("echo", ["$PWD"]);
            input.working_dir = Some(VirtualPath::Real(sandbox.path().into()));

            let ops = exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(input),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 0);

            if cfg!(unix) {
                assert_eq!(
                    exec.stdout.as_deref().unwrap().trim(),
                    sandbox.path().to_str().unwrap()
                );
            }

            // Custom dir
            sandbox.create_file("subdir/file", "");

            let mut input = ExecCommandInput::pipe("echo", ["$PWD"]);
            input.working_dir = Some(VirtualPath::Real(sandbox.path().join("subdir")));

            let ops = exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(input),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 0);

            if cfg!(unix) {
                assert_eq!(
                    exec.stdout.as_deref().unwrap().trim(),
                    sandbox.path().join("subdir").to_str().unwrap()
                );
            }

            // Custom dir via options
            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["$PWD"])),
                &ExecCommandOptions {
                    working_dir: Some(sandbox.path().join("subdir")),
                    ..Default::default()
                },
            )
            .await
            .unwrap();

            let exec = ops.first().unwrap().get_exec_output().unwrap();

            assert_eq!(exec.exit_code.unwrap(), 0);

            if cfg!(unix) {
                assert_eq!(
                    exec.stdout.as_deref().unwrap().trim(),
                    sandbox.path().join("subdir").to_str().unwrap()
                );
            }
        }

        #[tokio::test]
        async fn handles_on_exec() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let arg = Arc::new(Mutex::new(String::new()));
            let arg2 = arg.clone();

            exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["test"])),
                &ExecCommandOptions {
                    on_exec: Some(Arc::new(move |cmd, _| {
                        arg2.lock().unwrap().push_str(&cmd.command.args[0]);
                        Ok(())
                    })),
                    ..Default::default()
                },
            )
            .await
            .unwrap();

            assert_eq!(arg.lock().unwrap().as_str(), "test");
        }

        #[tokio::test]
        async fn will_retry_on_failure() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("exit", ["1"]))
                    .allow_failure()
                    .retry_count(2),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            assert_eq!(ops.len(), 3);
        }

        #[tokio::test]
        async fn doesnt_retry_on_success() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("exit", ["0"]))
                    .allow_failure()
                    .retry_count(2),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            assert_eq!(ops.len(), 1);
        }
    }

    mod exec_many {
        use super::*;

        fn get_order(ops: Vec<Operation>) -> Vec<String> {
            ops.into_iter()
                .map(|op| {
                    op.get_exec_output()
                        .unwrap()
                        .stdout
                        .as_ref()
                        .unwrap()
                        .trim()
                        .to_owned()
                })
                .collect()
        }

        #[tokio::test]
        async fn runs_serial_in_order() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let ops = exec_plugin_commands(
                "toolchain",
                ctx.clone(),
                vec![
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["1"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["2"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["3"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["4"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["5"])),
                ],
                ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            assert_eq!(get_order(ops), ["1", "2", "3", "4", "5"]);
        }

        // #[tokio::test]
        // async fn runs_parallel_in_any_order() {
        //     let (_, ws) = create_workspace();
        //     let ctx = Arc::new(ws.mock_app_context());

        //     let ops = exec_plugin_commands(
        //         "toolchain",
        //         ctx.clone(),
        //         vec![
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["1"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["2"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["3"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["4"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["5"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["6"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["7"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["8"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["9"])).parallel(),
        //             ExecCommand::new(ExecCommandInput::pipe("echo", ["10"])).parallel(),
        //         ],
        //         ExecCommandOptions::default(),
        //     )
        //     .await
        //     .unwrap();

        //     assert_ne!(
        //         get_order(ops),
        //         ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10"]
        //     );
        // }

        #[tokio::test]
        async fn runs_parallel_after_serial() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let ops = exec_plugin_commands(
                "toolchain",
                ctx.clone(),
                vec![
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["1"])).parallel(),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["2"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["3"])).parallel(),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["4"])),
                    ExecCommand::new(ExecCommandInput::pipe("echo", ["5"])).parallel(),
                ],
                ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let order = get_order(ops);

            assert_eq!(order[0], "2");
            assert_eq!(order[1], "4");
        }
    }

    mod cache {
        use super::*;

        fn get_hashes(root: &Path) -> Vec<PathBuf> {
            fs::read_dir(root.join(".moon/cache/hashes"))
                .unwrap()
                .map(|dir| dir.unwrap().path())
                .collect::<Vec<_>>()
        }

        #[tokio::test]
        async fn doesnt_create_cache_if_disabled() {
            let (sandbox, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            assert_eq!(get_hashes(sandbox.path()).len(), 0);
        }

        #[tokio::test]
        async fn creates_cache_if_enabled() {
            let (sandbox, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])).cache("key"),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            assert_eq!(get_hashes(sandbox.path()).len(), 1);
        }

        #[tokio::test]
        async fn skips_subsequent_execs() {
            let (_, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])).cache("key"),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])).cache("key"),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let op = ops.first().unwrap();

            assert_eq!(op.status, ActionStatus::Skipped);
        }

        #[tokio::test]
        async fn differentiates_between_keys() {
            let (sandbox, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            exec_plugin_command(
                ctx.clone(),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])).cache("key1"),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let ops = exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"])).cache("key2"),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let op = ops.first().unwrap();

            assert_ne!(op.status, ActionStatus::Skipped);

            assert_eq!(get_hashes(sandbox.path()).len(), 2);
        }

        #[tokio::test]
        async fn gathers_env_inputs() {
            let (sandbox, ws) = create_workspace();
            let ctx = Arc::new(ws.mock_app_context());

            let bag = GlobalEnvBag::instance();
            bag.set("EXISTS", "value");

            exec_plugin_command(
                ctx,
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"]))
                    .cache("key")
                    .inputs(vec![
                        CacheInput::EnvVar("EXISTS".into()),
                        CacheInput::EnvVar("MISSING".into()),
                    ]),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            bag.remove("EXISTS");

            let hashes = get_hashes(sandbox.path());
            let data: json::JsonValue = json::read_file(&hashes[0]).unwrap();

            assert_snapshot!(json::format(&data, true).unwrap());
        }

        #[tokio::test]
        async fn gathers_file_inputs_by_size() {
            let (sandbox, ws) = create_workspace();
            sandbox.create_file("input.txt", "test");

            exec_plugin_command(
                Arc::new(ws.mock_app_context()),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"]))
                    .cache("key")
                    .inputs(vec![
                        CacheInput::FileSize(VirtualPath::Real(sandbox.path().join("input.txt"))),
                        CacheInput::FileSize(VirtualPath::Real(sandbox.path().join("missing.txt"))),
                    ]),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let hashes = get_hashes(sandbox.path());
            let data: json::JsonValue = json::read_file(&hashes[0]).unwrap();

            assert_snapshot!(json::format(&data, true).unwrap());
        }

        #[tokio::test]
        async fn gathers_file_inputs_by_size_instead_of_hash_if_no_vcs() {
            let (sandbox, ws) = create_workspace();
            sandbox.create_file("input.txt", "test");

            exec_plugin_command(
                Arc::new(ws.mock_app_context()),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"]))
                    .cache("key")
                    .inputs(vec![
                        CacheInput::FileHash(VirtualPath::Real(sandbox.path().join("input.txt"))),
                        CacheInput::FileHash(VirtualPath::Real(sandbox.path().join("missing.txt"))),
                    ]),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let hashes = get_hashes(sandbox.path());
            let data: json::JsonValue = json::read_file(&hashes[0]).unwrap();

            assert_snapshot!(json::format(&data, true).unwrap());
        }

        #[tokio::test]
        async fn gathers_file_inputs_by_hash() {
            let (sandbox, ws) = create_workspace();
            sandbox.enable_git();
            sandbox.create_file("input.txt", "test");

            exec_plugin_command(
                Arc::new(ws.mock_app_context()),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"]))
                    .cache("key")
                    .inputs(vec![
                        CacheInput::FileHash(VirtualPath::Real(sandbox.path().join("input.txt"))),
                        CacheInput::FileHash(VirtualPath::Real(sandbox.path().join("missing.txt"))),
                    ]),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let hashes = get_hashes(sandbox.path());
            let data: json::JsonValue = json::read_file(&hashes[0]).unwrap();

            assert_snapshot!(json::format(&data, true).unwrap());
        }

        #[tokio::test]
        async fn gathers_file_inputs_by_timestamp() {
            let (sandbox, ws) = create_workspace();
            sandbox.create_file("input.txt", "test");

            exec_plugin_command(
                Arc::new(ws.mock_app_context()),
                &ExecCommand::new(ExecCommandInput::pipe("echo", ["cache"]))
                    .cache("key")
                    .inputs(vec![
                        CacheInput::FileTimestamp(VirtualPath::Real(
                            sandbox.path().join("input.txt"),
                        )),
                        CacheInput::FileTimestamp(VirtualPath::Real(
                            sandbox.path().join("missing.txt"),
                        )),
                    ]),
                &ExecCommandOptions::default(),
            )
            .await
            .unwrap();

            let hashes = get_hashes(sandbox.path());
            let data: json::JsonValue = json::read_file(&hashes[0]).unwrap();

            if let json::JsonValue::String(inner) = &data[0]["inputFiles"]["input.txt"] {
                assert!(inner.starts_with("timestamp:"));
            } else {
                panic!();
            }
        }
    }
}
