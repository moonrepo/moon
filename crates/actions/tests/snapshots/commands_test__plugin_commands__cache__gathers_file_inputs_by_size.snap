---
source: crates/actions/tests/commands_test.rs
expression: "json::format(&data, true).unwrap()"
---
[
  {
    "key": "key",
    "command": {
      "command": "echo",
      "args": [
        "cache"
      ],
      "set_executable": false,
      "stream": false
    },
    "inputFiles": {
      "input.txt": "size:4"
    }
  }
]
